server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java

```diff
@@ -90,6 +105,174 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+        final int simulateWorkTimeInMillis = 5;
+        int concurrentClients= 0;
+        int concurrentBulkRequests = 0;
+        int expectedExecutions= 0;
+        int maxBatchSize = 0;
+        int maxDocuments = 0;
+        int iterations = 0;
+        //find some randoms that allow this test to take under ~ 10 seconds
+        while (estimatedTimeForTest > 10_000) {
+            if (iterations++ > 1_000) {
+                fail("failed to find random values that allows test to run quickly"); //extremely unlikely
+            }
+            maxBatchSize = randomIntBetween(1, 100);
+            maxDocuments = randomIntBetween(maxBatchSize, 1_000_000);
+            concurrentClients = randomIntBetween(1, 20);
+            concurrentBulkRequests = randomIntBetween(0, 20);
+            expectedExecutions = maxDocuments / maxBatchSize;
+            estimatedTimeForTest = (expectedExecutions * simulateWorkTimeInMillis) /
+                Math.min(concurrentBulkRequests + 1, concurrentClients);
+        }
+        BulkResponse bulkResponse = new BulkResponse(new BulkItemResponse[]{new BulkItemResponse()}, 0);
+        AtomicInteger failureCount = new AtomicInteger(0);
+        AtomicInteger successCount = new AtomicInteger(0);
+        AtomicInteger requestCount = new AtomicInteger(0);
+        AtomicInteger docCount = new AtomicInteger(0);
+        BiConsumer<BulkRequest, ActionListener<BulkResponse>> consumer = (request, listener) ->
+        {
+            try {
+                Thread.sleep(simulateWorkTimeInMillis); //simulate work
+            } catch (InterruptedException e) {
+                failureCount.getAndIncrement();
+                logger.error("interrupted while sleeping. There is likely something wrong with this test!", e);
+            }
+            listener.onResponse(bulkResponse);
+        };
+        BulkProcessor bulkProcessor = new BulkProcessor(consumer, BackoffPolicy.noBackoff(),
+            countingListener(requestCount, successCount, failureCount, docCount),
+            concurrentBulkRequests, maxBatchSize, new ByteSizeValue(Integer.MAX_VALUE), null,
+            (command, delay, executor) -> null, () -> called.set(true), BulkRequest::new);
+
+        ExecutorService executorService = Executors.newFixedThreadPool(concurrentClients);
+
+        IndexRequest indexRequest = new IndexRequest();
+        String bulkRequest = "{ \"index\" : { \"_index\" : \"test\", \"_id\" : \"1\" } }\n" +  "{ \"field1\" : \"value1\" }\n";
+        BytesReference bytesReference =
+            BytesReference.fromByteBuffers(new ByteBuffer[]{ByteBuffer.wrap(bulkRequest.getBytes(StandardCharsets.UTF_8))});
+        List<Future> futures = new ArrayList<>();
+        for (final AtomicInteger i = new AtomicInteger(0); i.getAndIncrement() < maxDocuments;) {
+            //alternate between ways to add to the bulk processor
+            futures.add(executorService.submit(() -> {
+                if(i.get() % 2 == 0) {
```

Comment:
I am not sure this is the best way to test this. Ideally you want all of your threads ready to go prior to starting work.  Usually we use a cyclic barrier or countdown latch so that we know all threads are ready prior to execution of the concurrent tasks. An example use of a cyclic barrier would be https://github.com/elastic/elasticsearch/blob/0531987a0466fd7e4af367c76cf6b8911b4df51c/server/src/test/java/org/elasticsearch/common/cache/CacheTests.java#L843-L878.

Reply:
@jaymode - thanks for the pointer. I added countdown latch to ensure that all the tasks are submitted to the executor before doing any work. Let me know if this missed the mark. - https://github.com/elastic/elasticsearch/pull/41451/commits/a1e456c75a97ee28ff79695afc96c8503d34bf63